sensors/cmd/start.go

```diff
@@ -70,20 +45,194 @@ func Start() {
 	m := metrics.NewMetrics(sensor.Namespace)
 	go m.Run(ctx, fmt.Sprintf(":%d", common.SensorMetricsPort))
 
-	logger.Infow("starting sensor server", "version", argoevents.GetVersion())
-	sensorExecutionCtx := sensors.NewSensorContext(kubeClient, dynamicClient, sensor, busConfig, ebSubject, hostname, m)
-	if err := sensorExecutionCtx.Start(ctx); err != nil {
-		logger.Fatalw("failed to listen to events", zap.Error(err))
+	sensorCtx := sensors.NewSensorContext(scm.kubeClient, scm.dynamicClient, sensor, scm.eventBusConfig, scm.eventBusSubject, scm.hostname, m)
+	scm.C <- func() error { return sensorCtx.Start(ctx) }
+
+	return nil
+}
+
+func (scm *SensorContextManager) startLiveReload(ctx context.Context, path string) error {
+	// watch for sensor file changes
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+
+	err = watcher.Add(path)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		defer watcher.Close()
+		var m *metrics.Metrics
+
+		for {
+			sensor, err := scm.readSensorFromFile(path)
+			if err != nil {
+				scm.logger.Fatalw("failed to read sensor from file", zap.Error(err))
+			}
+
+			// logger
+			scm.logger = scm.logger.With("sensorName", sensor.Name)
+			for name, value := range sensor.Spec.LoggingFields {
+				scm.logger.With(name, value)
+			}
+
+			// start metrics once
+			if m == nil {
+				m = metrics.NewMetrics(sensor.Namespace)
+				go m.Run(ctx, fmt.Sprintf(":%d", common.SensorMetricsPort))
+			}
+
+			// create context
+			subCtx, cancel := context.WithCancel(ctx)
+			sensorCtx := sensors.NewSensorContext(scm.kubeClient, scm.dynamicClient, sensor, scm.eventBusConfig, scm.eventBusSubject, scm.hostname, m)
+			scm.C <- func() error { return sensorCtx.Start(subCtx) }
+
+			for {
+				select {
+				case event := <-watcher.Events:
+					if event.Op == fsnotify.Write || event.Op == fsnotify.Remove {
+						cancel()
+
+						// _ = watcher.Remove(path)
+						if err := watcher.Add(path); err != nil {
+							scm.logger.Fatalw("failed to add sensor watch", zap.Error(err))
+						}
+
+						break
+					}
+				case <-ctx.Done():
+					cancel()
+					return
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (scm *SensorContextManager) readSensorFromEnv() (*v1alpha1.Sensor, error) {
+	encodedSensorSpec, defined := os.LookupEnv(common.EnvVarSensorObject)
+	if !defined {
+		return nil, fmt.Errorf("required environment variable '%s' not defined", common.EnvVarSensorObject)
 	}
+	sensorSpec, err := base64.StdEncoding.DecodeString(encodedSensorSpec)
+	if err != nil {
+		return nil, err
+	}
+	sensor := &v1alpha1.Sensor{}
+	if err = json.Unmarshal(sensorSpec, sensor); err != nil {
+		return nil, err
+	}
+
+	return scm.verifySensor(sensor), nil
+}
+
+func (scm *SensorContextManager) readSensorFromFile(path string) (*v1alpha1.Sensor, error) {
+	sensorSpec, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	sensor := &v1alpha1.Sensor{}
+	if err = json.Unmarshal(sensorSpec, sensor); err != nil {
```

Comment:
I use `yaml.marshal. yaml` to get the encoded bytes for the configmap. im pretty sure json.unmarshal can unmarshal yaml too. just noting this

Reply:
I can switch to yaml